spec/client/ics-012-near-client/README.md

```diff
@@ -0,0 +1,550 @@
+                prevConsState = getPreviousConsensusState(header.getHeight())
+                clientState.frozenHeight = prevConsState.header.getHeight()
+            case Misbehaviour:
+                prevConsState = getPreviousConsensusState(clientMessage.header1.getHeight())
+                clientState.frozenHeight = prevConsState.header.getHeight()
+        }
+    }
+    set("clients/{clientMsg.identifier}/clientState", clientState)
+}
+```
+
+> As the slashing policy is NOT applicable in NEAR protocol for now, this section is only for references.
+
+### Upgrades
+
+The chain which this light client is tracking can elect to write a special pre-determined key in state to allow the light client to update its client state in preparation for an upgrade.
+
+As the client state change will be performed immediately, once the new client state information is written to the predetermined key, the client will no longer be able to follow blocks on the old chain, so it must upgrade promptly.
+
+```typescript
+function upgradeClientState(
+  clientState: ClientState,
+  newClientState: ClientState,
+  height: Height,
+  proof: CommitmentProof) {
+    // assert trusting period has not yet passed
+    assert(currentTimestamp() - clientState.latestTimestamp < clientState.trustingPeriod)
+    // check that the revision has been incremented
+    assert(newClientState.latestHeight > clientState.latestHeight)
+    // check proof of updated client state in state at predetermined commitment prefix and key
+    path = applyPrefix(clientState.upgradeCommitmentPrefix, clientState.upgradeKey)
+    // check that the client is at a sufficient height
+    assert(clientState.latestHeight >= height)
+    // fetch the previously verified commitment root & verify membership
+    root = get("clients/{clientMessage.identifier}/consensusStates/{height}")
+    // verify that the provided consensus state has been stored
+    assert(root.verifyMembership(path, newClientState, proof))
+    // update client state
+    clientState = newClientState
+    set("clients/{clientMessage.identifier}/clientState", clientState)
+}
+```
+
+### State verification functions
+
+The NEAR client state verification functions check a MPT (Merkle Patricia Tree) proof against a previously validated consensus state. The client should provide both membership verification and non-membership verification.
+
+```typescript
+function (ConsensusState) verifyMembership(
+  path: []byte,
+  value: []byte,
+  proof: [][]byte): bool {
+    // Check that the root in proof is one of the prevStateRoot of a chunk
+    assert(hash(proof[0]) in self.header.prevStateRootOfChunks)
+    // Check the value on the path is exactly the given value with proof data
+    // based on MPT construction algorithm
+}
+
+function verifyMembership(
+  clientState: ClientState,
+  height: Height,
+  delayTimePeriod: uint64,
+  delayBlockPeriod: uint64,
+  proof: CommitmentProof,
+  path: CommitmentPath,
+  value: []byte) {
```

Comment:
Check ICS 07 because this function should return and `Error`. Same for `verifyNonMembership`.

Reply:
OK. Will fix.